### PR TITLE
CSI-4229 delete snapshot 2.0

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: controller/array_action/array_mediator_svc.py [hunk 1]
```diff
@@ -720,12 +721,26 @@ def create_snapshot(self, volume_id, snapshot_name, space_efficiency, pool):
...
-        if not cli_volume or not cli_volume.FC_id:
-            raise array_errors.ObjectNotFoundError(snapshot_id)
-        self._delete_object(cli_volume, is_snapshot=True)
+        if self._is_flashcopy_2_0_supported() and not snapshot_id:
+            self._rmsnapshot(internal_snapshot_id)
+        else:
+            cli_volume = self._get_cli_volume_by_wwn(snapshot_id)
+            if not cli_volume or not cli_volume.FC_id:
+                raise array_errors.ObjectNotFoundError(snapshot_id)
+            self._delete_object(cli_volume, is_snapshot=True)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:743]: I think this should move to a `_delete_flashcopy_1_0_snapshot` - what do you think?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 2]
```diff
@@ -681,7 +682,25 @@ def test_delete_snapshot_with_stopfcmap_raise_error(self, mock_warning):
         mock_warning.return_value = False
         self.svc.client.svctask.stopfcmap.side_effect = [CLIFailureError('error')]
         with self.assertRaises(CLIFailureError):
-            self.svc.delete_snapshot("test_snapshot")
+            self.svc.delete_snapshot("test_snapshot", "internal_id")
+
+    def _prepare_mocks_for_delete_snapshot_2_0(self):
+        self.svc.client.svctask.addsnapshot = Mock()
+
+    def _test_delete_snapshot_2_0_rmsnapshot_cli_failure_error(self, error_message_id, expected_error):
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:690]:
```suggestion
    def _test_delete_snapshot_flashcopy_2_0_rmsnapshot_cli_failure_error(self, error_message_id, expected_error):
```
or if we don't want the storage feature name in the test functions name:
```suggestion
    def _test_delete_snapshot_rmsnapshot_cli_failure_error(self, error_message_id, expected_error):
```

Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: done

**Review by oriyarde — APPROVED** [review]
LGTM

**Review by matancarmeli7 — APPROVED** [review]
LGTM

File: controller/array_action/array_mediator_svc.py [hunk 3]
```diff
@@ -720,12 +723,23 @@ def create_snapshot(self, volume_id, snapshot_name, space_efficiency, pool):
...
+    def _is_addsnapshot_supported(self):
+        return hasattr(self.client.svctask, "addsnapshot")
+
+    def _rmsnapshot(self, internal_snapshot_id):
+        try:
+            self.client.svctask.rmsnapshot(snapshotid=internal_snapshot_id)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if SNAPSHOT_NOT_EXIST in ex.my_message:
+                raise array_errors.ObjectNotFoundError(internal_snapshot_id)
+            raise ex
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:735]: should not fail [if exception is just a warning](https://github.com/IBM/ibm-block-csi-driver/pull/538#discussion_r917422095)

**Pull request merged**